File: src/examples/USExtractExample.cs
namespace Examples
{
	using System;
    using System.IO;
    using System.Net;
    using System.Threading.Tasks;
    using SmartyStreets;
	using SmartyStreets.USExtractApi;

	internal static class USExtractExample
	{
		public static async Task Run()
		{
            // specifies the TLS protocoll to use - this is TLS 1.2
            const SecurityProtocolType tlsProtocol1_2 = (SecurityProtocolType)3072;

            // We recommend storing your secret keys in environment variables.
            var authId = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID");
			var authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN");
			ServicePointManager.SecurityProtocol = tlsProtocol1_2;

			var client = new ClientBuilder(authId, authToken).BuildUsExtractApiClient();
			var text = "Here is some text.\r\nMy address is 3785 Las Vegs Av." +
			           "\r\nLos Vegas, Nevada." +
			           "\r\nMeet me at 1 Rosedale Baltimore Maryland, not at 123 Phony Street, Boise Idaho.";
			
			// Documentation for input fields can be found at:
			// https://smartystreets.com/docs/cloud/us-extract-api#http-request-input-fields
			
			var lookup = new Lookup(text)
			{
				IsAggressive = true,
				AddressesHaveLineBreaks = false,
				AddressesPerLine = 1
			};

			//uncomment the line below to add a custom parameter
			//lookup.AddCustomParameter("addr_line_breaks", "false");
			
            try
            {
                await client.Send(lookup);
            }
            catch (SmartyException ex)
            {
                Console.WriteLine(ex.Message);
                Console.WriteLine(ex.StackTrace);
				return;
            }
            catch (IOException ex)
            {
                Console.WriteLine(ex.StackTrace);
				return;
            }

			var result = lookup.Result;

            if (result.Metadata == null)
            {
                Console.WriteLine("No candidates. This means the address is not valid.");
                return;
            }

            var metadata = result.Metadata;
			Console.WriteLine("Found " + metadata.AddressCount + " addresses.");
			Console.WriteLine(metadata.VerifiedCount + " of them were valid.");
			Console.WriteLine();

			var addresses = result.Addresses;

			Console.WriteLine("Addresses: \r\n**********************\r\n");
			foreach (var address in addresses)
			{
				Console.WriteLine("\"" + address.Text + "\"\n");
				Console.WriteLine("Verified? " + address.Verified);
				if (address.Candidates.Length > 0)
				{
					Console.WriteLine("\nMatches:");

					foreach (var candidate in address.Candidates)
					{
						Console.WriteLine(candidate.DeliveryLine1);
						Console.WriteLine(candidate.LastLine);
						Console.WriteLine();
					}
				}
				else
				{
					Console.WriteLine();
				}

				Console.WriteLine("**********************\n");
			}
		}
	}
}

File: src/vbexamples/USExtractExample.vb
Imports SmartyStreets.USExtractApi
Imports SmartyStreets
Imports System.IO
Imports System.Formats

Module USExtractExample

    Dim authID = Environment.GetEnvironmentVariable("SMARTY_AUTH_ID")
    Dim authToken = Environment.GetEnvironmentVariable("SMARTY_AUTH_TOKEN")

    Dim client = New ClientBuilder(authID, authToken).WithLicense(New List(Of String) From {"us-core-cloud"}).BuildUsExtractApiClient()

    Dim text = "Here is some text." + Environment.NewLine + "My address is 3785 Las Vegs Av." + Environment.NewLine +
                       "Los Vegas, Nevada." + Environment.NewLine +
                       "Meet me at 1 Rosedale Baltimore Maryland, not at 123 Phony Street, Boise Idaho."

    Sub USExtractExample()

        Dim lookup As New Lookup()
        With lookup
            .Text = text
            .IsAggressive = True
            .AddressesHaveLineBreaks = False
            .AddressesPerLine = 1
        End With

        Console.WriteLine("*******************************************************")
        Console.WriteLine()

        Try
            client.Send(lookup)
        Catch ex As SmartyException
            Console.WriteLine(ex.Message)
            Console.WriteLine(ex.StackTrace)
        Catch ex As IOException
            Console.WriteLine(ex.StackTrace)
        Catch ex As Exception
            Console.WriteLine(ex.Message)
            Console.WriteLine(ex.StackTrace)
        End Try

        Dim result = lookup.Result
        Dim metadata = result.Metadata
        Dim addressCount = metadata.AddressCount
        Dim verifiedCount = metadata.VerifiedCount

        Console.WriteLine("Original lookup: " + lookup.Text + Environment.NewLine())

        Console.WriteLine("Extracted " + CStr(addressCount) + " address" + If(addressCount = 1, ".", "es."))
        Console.WriteLine(CStr(verifiedCount) + " of them " + If(addressCount = 1, "was", "were") + " valid.")
        Console.WriteLine()

        Dim addresses = result.Addresses

        If addressCount > 0 Then
            Console.WriteLine("Extracted address" + If(addressCount = 1, ":", "es:"))
        End If

        For Each address In addresses
            Console.WriteLine("""" + address.Text + """")
            Console.WriteLine("Verified? " + CStr(address.Verified) + Environment.NewLine)

            Dim candidatesCount = address.Candidates.Length

            If (candidatesCount > 0) Then
                Console.WriteLine("Found " + CStr(address.Candidates.Length) + " address match" + If(candidatesCount = 1, ":", "es:") + Environment.NewLine())

                For Each candidate In address.Candidates
                    Console.WriteLine(candidate.DeliveryLine1)
                    Console.WriteLine(candidate.LastLine)
                    Console.WriteLine()
                Next
            Else
                Console.WriteLine()
            End If

        Next

    End Sub

End Module
